## Explore and cluster the neighborhoods in Toronto

Scarping a Wikipedia webpage with BeautifulSoup for Toronto Nieghborhood data 

##  Import and install required libraries/modules 

In [1]:
import pandas as pd 
!conda install -c conda-forge beautifulsoup4 --yes         #install beautifulsoup
from bs4 import BeautifulSoup              #for sacraping a webpage
from pandas import DataFrame       #for convering a list to dataframe
import requests        #make a request to get the webpage/url               
print('import and install finished')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - beautifulsoup4


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    beautifulsoup4-4.9.1       |   py36h9f0ad1d_0         163 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.6 MB

The following NEW packages will be INSTALLED:

    python_abi:      3.6-1_cp36m       conda-forge

The following packages will be UPDATED:

    beautifulsoup4:  4.7.1-py36_1                

##  Request, Scrape (with BeautifulSoup) , Read (with pandas) and convert to Dataframe and name it Toro 

In [2]:
res = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
soup = BeautifulSoup(res.content,'html.parser')    #scrape & parse 
table = soup.find_all('table')[0]    
df = pd.read_html(str(table))[0]    
#made three lists from df for DataFrame
Postal_code = df["Postal Code"].tolist()
Borough = df["Borough"].tolist()
Neighborhood = df["Neighborhood"].tolist()
#make a dictionary for DataFrame
d = {'PostalCodes':Postal_code,'Borough':Borough, 'Neighborhood':Neighborhood}  
#made the DataFrame
Toro = pd.DataFrame(d)
Toro.head()

PostalCodes           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

##  Remove cells/rows with "Not assigned" Borough

In [4]:
# Get names of indices for which column Borough has value Not assigned
indexNames = Toro[ Toro['Borough'] == 'Not assigned' ].index
indexNames
# Delete these row indices from dataFrame
Toro.drop(indexNames , inplace=True)  #.reset_index(drop=True)
Toro.head()

PostalCodes           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

## Reset index and drop old index

In [7]:
print(Toro.index)

Int64Index([  2,   3,   4,   5,   6,   8,   9,  11,  12,  13,
            ...
            151, 152, 153, 156, 157, 160, 165, 168, 169, 178],
           dtype='int64', length=103)


In [7]:
Toro.reset_index(drop=True, inplace=True)

In [8]:
Toro.head()

PostalCodes           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [9]:
Toro.shape

(103, 3)

## Check if there is any Borough with "Not assigned" Neighborhood

In [10]:
df = Toro
df.loc[df['Neighborhood'] != 'Not assigned', 'Assigned'] = 'Yes'
df.describe() #check if any Neiborhood has Not assighned value

PostalCodes     Borough Neighborhood Assigned
count          103         103          103      103
unique         103          10           99        1
top            M4Y  North York    Downsview      Yes
freq             1          24            4      103

## From the above result all 103 Neighborhoods have been Assigned  

In [11]:
Toro.head()

PostalCodes           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

  Assigned  
0      Yes  
1      Yes  
2      Yes  
3      Yes  
4      Yes

In [12]:
Toro.drop('Assigned',axis = 1, inplace=True)

In [13]:
Toro.head()

PostalCodes           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [14]:
Toro.shape

(103, 3)

## Part 2 - adding latitude and longtide 

The geocoder provided did not work. We will read the csv file provided by the link. 

In [15]:
import pandas as pd

path = "https://cocl.us/Geospatial_data"
df2 = pd.read_csv(path)
df2.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## The csv data seems to be sorted by Postal Code. Let's sort our dataframe to match it.

In [16]:
Toro.sort_values('PostalCodes',inplace=True)
Toro.head()

PostalCodes      Borough                            Neighborhood
6          M1B  Scarborough                          Malvern, Rouge
12         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
18         M1E  Scarborough       Guildwood, Morningside, West Hill
22         M1G  Scarborough                                  Woburn
26         M1H  Scarborough                               Cedarbrae

## A quick scan of the above sorted dataframe matches with csv file. Next, reseting and droping old index.

In [17]:
Toro.reset_index(drop=True, inplace=True)

In [18]:
Toro.head()

PostalCodes      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

##  Now that the two dataframes match, let's concatenate them

In [19]:
Toro_data = pd.concat([Toro, df2], axis=1, sort=False)
Toro_data.head()

PostalCodes      Borough                            Neighborhood  \
0         M1B  Scarborough                          Malvern, Rouge   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek   
2         M1E  Scarborough       Guildwood, Morningside, West Hill   
3         M1G  Scarborough                                  Woburn   
4         M1H  Scarborough                               Cedarbrae   

  Postal Code   Latitude  Longitude  
0         M1B  43.806686 -79.194353  
1         M1C  43.784535 -79.160497  
2         M1E  43.763573 -79.188711  
3         M1G  43.770992 -79.216917  
4         M1H  43.773136 -79.239476

## Let's drop the extra column Postal Code

In [20]:
Toro_data.drop('Postal Code', axis = 1, inplace=True)
Toro_data.head()

PostalCodes      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [21]:
Toro_data.head(12)

PostalCodes      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   

                                         Neighborhood   Latitude  Longitude  
0                                      Malvern, Rouge  43.806686 -79.194353  
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                           Cedarbrae  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029  
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577  
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476  
9                         Birch Cliff, Cliffside West  43.692657 -79.264848  
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304  
11                                  Wexford, Maryvale  43.750072 -79.295849

In [22]:
Toro_data.shape

(103, 5)

In [23]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(Toro_data['Borough'].unique()),
        Toro_data.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


## install and import - geopy, matplotlib and folium 

In [25]:
#import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    branca-0.4.1               |             py_0          26 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         713 KB

The following NEW packages will be INSTALLED:

    altair:  4.1.0-py_1 conda-forge
    branca:  0.4.1-py_0 conda-forge
    folium:  0.5.0-py_0 conda-forge
    vincent: 0.4.4-py_1 conda-forge


bran

## Use geopy library to get the latitude and longitude values of Toronto City with user agent to_explorer

In [26]:
address = 'Toronto, ON, Canada'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


## Map of Toronto with neighborhoods superimposed on top

In [28]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(Toro_data['Latitude'], Toro_data['Longitude'], Toro_data['Borough'], Toro_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Utilizing the Foursquare API to explore the neighborhoods

In [29]:
CLIENT_ID = 'AS44E1HIUXF1EBWBYWFENQB3F3BOW01C1LMJ1XMU0H0WE4SP' # your Foursquare ID
CLIENT_SECRET = 'STU4NMJZ2Y5NAC1T42OCGGSP2E2BEWUJXOCLT5WUHZQ0Y13U' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: AS44E1HIUXF1EBWBYWFENQB3F3BOW01C1LMJ1XMU0H0WE4SP
CLIENT_SECRET:STU4NMJZ2Y5NAC1T42OCGGSP2E2BEWUJXOCLT5WUHZQ0Y13U


## Let's start with Downtown Toronto.

In [32]:
downtown_data = Toro_data[Toro_data['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_data.head()

PostalCodes           Borough                 Neighborhood   Latitude  \
0         M4W  Downtown Toronto                     Rosedale  43.679563   
1         M4X  Downtown Toronto  St. James Town, Cabbagetown  43.667967   
2         M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3         M5A  Downtown Toronto    Regent Park, Harbourfront  43.654260   
4         M5B  Downtown Toronto     Garden District, Ryerson  43.657162   

   Longitude  
0 -79.377529  
1 -79.367675  
2 -79.383160  
3 -79.360636  
4 -79.378937

## Downtown Toronto latitude and longitude

In [33]:
address = 'Downtown Toronto, Toronto, ON, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronoto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronoto are 43.6563221, -79.3809161.


## Map of Downtown Toronto 

In [37]:
# create map of Manhattan using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, label in zip(downtown_data['Latitude'], downtown_data['Longitude'], downtown_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

## Foursquare credintials 

In [38]:
CLIENT_ID = 'AS44E1HIUXF1EBWBYWFENQB3F3BOW01C1LMJ1XMU0H0WE4SP' # your Foursquare ID
CLIENT_SECRET = 'STU4NMJZ2Y5NAC1T42OCGGSP2E2BEWUJXOCLT5WUHZQ0Y13U' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: AS44E1HIUXF1EBWBYWFENQB3F3BOW01C1LMJ1XMU0H0WE4SP
CLIENT_SECRET:STU4NMJZ2Y5NAC1T42OCGGSP2E2BEWUJXOCLT5WUHZQ0Y13U


## Let's look at the first neighborhood in Downtown Toronto

In [39]:
downtown_data.loc[0, 'Neighborhood']

'Rosedale'

## Get latitude and longtide for the first neighborhood

In [40]:
neighborhood_latitude = downtown_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = downtown_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = downtown_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rosedale are 43.6795626, -79.37752940000001.


## Let's get the url but change the limit to 1000 meters for more results

In [43]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=AS44E1HIUXF1EBWBYWFENQB3F3BOW01C1LMJ1XMU0H0WE4SP&client_secret=STU4NMJZ2Y5NAC1T42OCGGSP2E2BEWUJXOCLT5WUHZQ0Y13U&v=20180605&ll=43.6795626,-79.37752940000001&radius=1000&limit=100'

In [44]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eff57b83cd8c443738a2763'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Rosedale',
  'headerFullLocation': 'Rosedale, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 21,
  'suggestedBounds': {'ne': {'lat': 43.68856260900001,
    'lng': -79.36510816548741},
   'sw': {'lat': 43.670562590999985, 'lng': -79.38995063451262}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4adcb343f964a520e32e21e3',
       'name': 'Summerhill Market',
       'location': {'address': '446 Summerhill Ave',
        'crossStreet': 'btwn. MacLennan Ave. and Glen Rd.',
        'lat': 43.68626482142425,
        'lng': -79.37545823237794,
      

## Function that extracts the category of the venue

In [45]:

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

## Clean the json and structure it into a pandas dataframe.

In [47]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name          categories        lat        lng
0         Summerhill Market       Grocery Store  43.686265 -79.375458
1               Black Camel           BBQ Joint  43.677016 -79.389367
2        Craigleigh Gardens                Park  43.678099 -79.371586
3  Toronto Lawn Tennis Club  Athletics & Sports  43.680667 -79.388559
4               Pie Squared            Pie Shop  43.672143 -79.377856

In [48]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

21 venues were returned by Foursquare.


## Exploring Neighborhoods in Downtown Toronto

In [49]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

## Use function above on each neighborhood and create a new dataframe called downtown_venues

In [51]:
# type your answer here
downtown_venues = getNearbyVenues(names=downtown_data['Neighborhood'],
                                   latitudes=downtown_data['Latitude'],
                                   longitudes=downtown_data['Longitude']
                                  )



Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Queen's Park, Ontario Provincial Government


In [52]:
print(downtown_venues.shape)
downtown_venues.head()

(1217, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                     Rosedale              43.679563              -79.377529   
1                     Rosedale              43.679563              -79.377529   
2                     Rosedale              43.679563              -79.377529   
3                     Rosedale              43.679563              -79.377529   
4  St. James Town, Cabbagetown              43.667967              -79.367675   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0         Rosedale Park       43.682328       -79.378934     Playground  
1          Whitney Park       43.682036       -79.373788           Park  
2  Alex Murray Parkette       43.678300       -79.382773           Park  
3        Milkman's Lane       43.676352       -79.373842          Trail  
4           Cranberries       43.667843       -79.369407          Diner

## Number of venues returned for each neighborhood

In [53]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              60   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
St. James Town                                                         79   
St. James Town, Cabbagetown                                            42   
Stn A PO Boxes                                                         97   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               60   
Queen's Park, Ontario Provincial Government                             32   
Regent Park, Harbourfront                                               44   
Richmond, Adelaide, King                                                94   
Rosedale                                                                 4   
St. James Town                                                          79   
St. James Town, Cabbagetown                                             42   
Stn A PO Boxes                                                          97   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          35   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            58              58   
CN Tower, King and Spadina, Railway Lands, Harb...     15              15   
Central Bay Street                                     65              65   
Christie                                               17              17   
Church and Wellesley                                   75              75   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

In [54]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 206 uniques categories.


## Analyzing Each Neighborhood

In [56]:
# hotspots
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  \
0                 0                   0          0           0           0   
1                 0                   0          0           0           0   
2                 0                   0          0           0           0   
3                 0                   0          0           0           0   
4                 0                   0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Butcher  \
0        0                0         0             0              0        0   
1        0                0         0             0              0        0   
2        0                0         0             0              0        0   
3        0                0         0             0              0        0   
4        0                0         0             0              0        0   

   Café  Camera Store  Candy Store  Caribbean Restaurant  Cheese Shop  \
0     0             0            0                     0            0   
1     0             0            0                     0            0   
2     0             0            0                     0            0   
3     0             0            0                     0            0   
4     0             0            0                     0    

## top 5 most common venues by neighborhood

In [57]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                  Central Bay Street     0.015385   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.026667   
5                      Commerce Court, Victoria Hotel     0.000000   
6              First Canadian Place, Underground city     0.000000   
7                            Garden District, Ryerson     0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000   
10        Queen's Park, Ontario Provincial Government     0.031250   
11                          Regent Park, Harbourfront     0.022727   
12                           Richmond, Adelaide, King     0.000000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                        St. James Town, Cabbagetown     0.000000   
16                                     Stn A PO Boxes     0.010309   
17           Toronto Dominion Centre, Design Exchange     0.000000   
18                     University of Toronto, Harbord     0.028571   

    Afghan Restaurant   Airport  Airport Food Court  Airport Lounge  \
0            0.000000  0.000000            0.000000        0.000000   
1            0.000000  0.066667            0.066667        0.133333   
2            0.000000  0.000000            0.000000        0.000000   
3            0.000000  0.000000            0.000000        0.000000   
4            0.013333  0.000000            0.000000        0.000000   
5            0.000000  0.000000            0.000000        0.000000   
6            0.000000  0.000000            0.000000        0.000000   
7            0.000000  0.000000            0.000000        0.000000   
8            0.000000  0.000000            0.000000        0.000000   
9            0.000000  0.000000            0.000000        0.000000   
10           0.000000  0.000000            0.000000        0.000000   
11           0.000000  0.000000            0.000000        0.000000   
12           0.000000  0.000000            0.000000        0.000000   
13           0.000000  0.000000            0.000000        0.000000   
14           0.000000  0.000000            0.000000        0.000000   
15           0.000000  0.000000            0.000000        0.000000   
16           0.000000  0.000000            0.000000        0.000000   
17           0.000000  0.000000            0.000000        0.000000   
18           0.000000  0.000000            0.000000        0.000000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0               0.0          0.000000             0.000000      0.000000   
1               0.2          0.066667             0.000000      0.000000   
2               0.0          0.000000             0.000000      0.000000   
3               0.0          0.000000             0.000000      0.000000   
4               0.0          0.000000             0.013333      0.000000   
5               0.0          0.000000             0.040000      0.000000   
6               0.0          0.000000             0.030000      0.000000   
7               0.0          0.000000             0.000000      0.000000   
8               0.0          0.000000             0.000000      0.000000   
9               0.0          0.000000             0.000000      0.000000   
10              0.0          0.000000             0.000000      0.000000   
11              0.0          0.000000             0.000000      0.022727   
12              0.0          0.000000             0.010638      0.000000   
13              0.0          0.000000             0.000000      0.000000   
14              0.0          0.000000             0.037975      0.000000   
15 

In [58]:
num_top_venues = 5

for hood in downtown_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_grouped[downtown_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2                Café  0.03
3  Seafood Restaurant  0.03
4         Cheese Shop  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
             venue  freq
0  Airport Service  0.20
1   Airport Lounge  0.13
2    Boat or Ferry  0.07
3  Harbor / Marina  0.07
4         Boutique  0.07


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.17
1       Sandwich Place  0.06
2   Italian Restaurant  0.06
3                 Café  0.05
4  Japanese Restaurant  0.05


----Christie----
                venue  freq
0       Grocery Store  0.24
1                Café  0.18
2                Park  0.12
3  Italian Restaurant  0.06
4               Diner  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.08
1     Sushi Restaurant  0.07
2  Japanese Restaurant  0.05
3       

## puting it in pandas frame

In [59]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [61]:
import numpy as np

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop            Restaurant   
1        Airport Lounge      Sculpture Garden         Boat or Ferry   
2        Sandwich Place    Italian Restaurant   Japanese Restaurant   
3                  Café                  Park           Coffee Shop   
4      Sushi Restaurant   Japanese Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
0              Beer Bar    Seafood Restaurant                    Bakery   
1   Rental Car Location              Boutique                       Bar   
2                  Café       Thai Restaurant                       Bar   
3             Nightclub                 Diner        Italian Restaurant   
4               Gay Bar           Yoga Studio  Mediterranean Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café                 Beach          Shopping Mall  
1       Harbor / Marina           Coffee Shop       Airport Terminal  
2       Bubble Tea Shop          Burger Joint       Department Store  
3           Candy Store    Athletics & Sports             Restaurant  
4           Men's Store                 Hotel        Bubble Tea Shop

##  Clustering Neighborhoods

In [ ]:
## k-means to cluster the neighborhoods into 5 clusters

In [62]:
# set number of clusters
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 2, 4, 3, 0, 0, 0, 0, 0, 0], dtype=int32)

## new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [63]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = downtown_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

downtown_merged.head() # check the last columns!

PostalCodes           Borough                 Neighborhood   Latitude  \
0         M4W  Downtown Toronto                     Rosedale  43.679563   
1         M4X  Downtown Toronto  St. James Town, Cabbagetown  43.667967   
2         M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3         M5A  Downtown Toronto    Regent Park, Harbourfront  43.654260   
4         M5B  Downtown Toronto     Garden District, Ryerson  43.657162   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529               1                  Park            Playground   
1 -79.367675               0           Coffee Shop                Bakery   
2 -79.383160               0           Coffee Shop      Sushi Restaurant   
3 -79.360636               0           Coffee Shop                Bakery   
4 -79.378937               0        Clothing Store           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                 Trail              Creperie           Doner Restaurant   
1           Pizza Place            Restaurant                       Café   
2   Japanese Restaurant            Restaurant                    Gay Bar   
3                   Pub                  Park             Breakfast Spot   
4        Cosmetics Shop       Bubble Tea Shop  Middle Eastern Restaurant   

  6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
0               Dog Run       Distribution Center        Discount Store   
1    Italian Restaurant                       Pub             Gastropub   
2           Yoga Studio  Mediterranean Restaurant           Men's Store   
3                  Café                   Theater           Yoga Studio   
4                  Café       Japanese Restaurant      Ramen Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                 Diner           Dessert Shop  
1            Playground              Pet Store  
2                 Hotel        Bubble Tea Shop  
3          Dessert Shop             Restaurant  
4             Bookstore                 Bakery

## Let's visiualize this

In [64]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighborhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Clusters

## Cluster 1

In [65]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0        Clothing Store   
5   Downtown Toronto               0                  Café   
6   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0                  Café   
13  Downtown Toronto               0                  Café   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1                 Bakery           Pizza Place             Restaurant   
2       Sushi Restaurant   Japanese Restaurant             Restaurant   
3                 Bakery                   Pub                   Park   
4            Coffee Shop        Cosmetics Shop        Bubble Tea Shop   
5            Coffee Shop             Gastropub             Restaurant   
6           Cocktail Bar           Cheese Shop             Restaurant   
8                   Café            Restaurant                    Gym   
9               Aquarium                  Café                  Hotel   
10                 Hotel                  Café             Restaurant   
11            Restaurant                  Café                  Hotel   
12   Japanese Restaurant            Restaurant                    Bar   
13                Bakery    Mexican Restaurant  Vietnamese Restaurant   
15    Seafood Restaurant                  Café                  Hotel   
16                  Café            Restaurant                  Hotel   

        5th Most Common Venue          6th Most Common Venue  \
1                        Café             Italian Restaurant   
2                     Gay Bar                    Yoga Studio   
3              Breakfast Spot                           Café   
4   Middle Eastern Restaurant                           Café   
5         American Restaurant                   Cocktail Bar   
6                    Beer Bar             Seafood Restaurant   
8             Thai Restaurant                          Hotel   
9                  Restaurant             Italian Restaurant   
10        American Restaurant                    Salad Place   
11        American Restaurant                            Gym   
12         Italian Restaurant                      Bookstore   
13                Coffee Shop  Vegetarian / Vegan Restaurant   
15         Italian Restaurant                     Restaurant   
16                        Gym            Japanese Restaurant   

       7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1                        Pub             Gastropub            Playground   
2   Mediterranean Restaurant           Men's Store                 Hotel   
3                    Theater           Yoga Studio          Dessert Shop   
4        Japanese Restaurant      Ramen Restaurant             Bookstore   
5                        Gym        Clothing Store                 Hotel   
6                     Bakery                  Café                 Beach   
8              Deli / Bodega        Clothing Store           Salad Place   
9        Sporting Goods Shop        Scenic Lookout   Fried Chicken Joint   
10       Japanese Restaurant    Italian Restaurant    Seafood Restaurant   
11       Japanese Restaurant    Italian Restaurant    Seafood Restaurant   
12                    Bakery              Beer Bar            Beer Store   
13              Dessert Shop           Pizza Place           Gaming Cafe   
15              Cocktail Bar         

## Cluster 2

In [66]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0  Downtown Toronto               1                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Playground                 Trail              Creperie   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Doner Restaurant               Dog Run   Distribution Center   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Discount Store                 Diner           Dessert Shop

## Cluster 3 - Again more than one would have been ideal

In [67]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               2       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14        Airport Lounge      Sculpture Garden         Boat or Ferry   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14   Rental Car Location              Boutique                   Bar   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14       Harbor / Marina           Coffee Shop       Airport Terminal

## cluster 4 - more than one would have been better

In [68]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               3         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Café                  Park           Coffee Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17             Nightclub                 Diner    Italian Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17           Candy Store    Athletics & Sports             Restaurant

## cluster 5 - more than one would have been better

In [69]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
7   Downtown Toronto               4           Coffee Shop   
18  Downtown Toronto               4           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7         Sandwich Place    Italian Restaurant   Japanese Restaurant   
18                 Diner      Sushi Restaurant                   Gym   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
7                   Café       Thai Restaurant                   Bar   
18                  Park    Mexican Restaurant    Italian Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
7        Bubble Tea Shop          Burger Joint       Department Store  
18            Hobby Shop   Fried Chicken Joint    Distribution Center